File: packages/react-native-nitro-svg-image/android/src/main/java/com/margelo/nitro/svg/image/HybridSVGImageFactory.kt
package com.margelo.nitro.svg.image

import android.graphics.Bitmap
import android.graphics.Canvas
import com.caverock.androidsvg.SVG
import com.margelo.nitro.image.HybridImage
import com.margelo.nitro.image.HybridImageSpec

class HybridSVGImageFactory : HybridSVGImageFactorySpec() {
    override fun stringToImage(url: String, options: AsyncImageLoadOptions?): HybridImageSpec {
        // Use url as SVG string
        val svgImage = SVG.getFromString(url)
        val width = svgImage.getDocumentWidth().toInt()
        val height = svgImage.getDocumentHeight().toInt()
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        svgImage.renderToCanvas(canvas)
        return HybridImage(bitmap)
    }
}

File: packages/react-native-nitro-svg-image/ios/HybridSVGImageFactory.swift
//
//  HybridSVGImageFactory.swift
//  react-native-nitro-svg-image
//
//  Created by Marc Rousavy on 10.06.25.
//

import Foundation
import NitroImage
import NitroModules
import SDWebImage
import SVGKit

private class HybridImage: HybridImageSpec, NativeImage {
    let uiImage: UIImage
    init(uiImage: UIImage) {
        self.uiImage = uiImage
        super.init()
    }
}

class HybridSVGImageFactory: HybridSVGImageFactorySpec {
    func stringToImage(url urlString: String, options: AsyncImageLoadOptions?) throws
        -> any HybridImageSpec
    {
        // Use urlString as SVG string with SVGKit
        guard let svgImage = SVGKImage(data: urlString.data(using: .utf8)) else {
            throw RuntimeError.error(withMessage: "Failed to parse SVG string")
        }

        // Convert SVGKit image to UIImage
        guard let uiImage = svgImage.uiImage else {
            throw RuntimeError.error(withMessage: "Failed to convert SVGKit image to UIImage")
        }

        // Return as HybridImage
        return HybridImage(uiImage: uiImage)
    }
}
